src/main/java/com/faforever/api/user/UserService.java

```diff
@@ -124,4 +140,104 @@ void activate(String token) {
     log.debug("User has been activated: {}", user);
     userRepository.save(user);
   }
+
+  void changePassword(String currentPassword, String newPassword, User user) {
+    if (!Objects.equals(user.getPassword(), passwordEncoder.encode(currentPassword))) {
+      throw new ApiException(new Error(ErrorCode.PASSWORD_CHANGE_FAILED));
+    }
+
+    user.setPassword(passwordEncoder.encode(newPassword));
+    userRepository.save(user);
+  }
+
+  void changeLogin(String newLogin, User user) {
+    validateUsername(newLogin);
+
+    int minDaysBetweenChange = properties.getUser().getMinimumDaysBetweenUsernameChange();
+    nameRecordRepository.getDaysSinceLastNewRecord(user.getId(), minDaysBetweenChange)
+      .ifPresent(daysSinceLastRecord -> {
+        throw new ApiException(new Error(ErrorCode.USERNAME_CHANGE_TOO_EARLY, minDaysBetweenChange - daysSinceLastRecord));
+      });
+
+    int usernameReservationTimeInMonths = properties.getUser().getUsernameReservationTimeInMonths();
+    nameRecordRepository.getLastUsernameOwnerWithinMonths(newLogin, usernameReservationTimeInMonths)
+      .ifPresent(reservedByUserId -> {
+        if (reservedByUserId != user.getId()) {
```

Comment:
This is an object comparison that needs to be done using `equals()`

Reply:
Why? We are comparing to ints.

Reply:
No, we are comparing two `Integer`. `==` will thus compare references.

Reply:
`reservedByUserId` is an Integer, while user.getId() returns `int`
If I  use `reservedByUserId.intValue()` IntelliJ starts complaining, so I leave it as is.